soroban-sdk/src/bytes.rs

```diff
@@ -1233,4 +1233,240 @@ mod test {
         let arr_bin: BytesN<3> = bin.clone().try_into().unwrap();
         assert_eq!(format!("{:?}", arr_bin), "BytesN<3>(10, 20, 30)");
     }
+
+    #[test]
+    fn test_is_empty() {
+        let env = Env::default();
+        let mut bin = Bytes::new(&env);
+        assert_eq!(bin.is_empty(), true);
+        bin.push(10);
+        assert_eq!(bin.is_empty(), false);
+    }
+
+    #[test]
+    fn test_first() {
+        let env = Env::default();
+        let mut bin = bytes![&env, [1, 2, 3, 4]];
+
+        assert_eq!(bin.first(), Some(1));
+        bin.remove(0);
+        assert_eq!(bin.first(), Some(2));
+
+        // first on empty bytes
+        let bin = bytes![&env];
+        assert_eq!(bin.first(), None);
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_first_unchecked() {
+        let env = Env::default();
+        let mut bin = bytes![&env, [1, 2, 3, 4]];
+
+        assert_eq!(bin.first_unchecked(), 1);
+        bin.remove(0);
+        assert_eq!(bin.first_unchecked(), 2);
+
+        // first on empty bytes should panic
+        let bin = bytes![&env];
+        bin.first_unchecked();
```

Comment:
This test is somewhat brittle because if any part of the test panics it'll pass, even though the only case we expect to panic is the last line. It would be ideal if we could replace the `#[should_panic]` with wrapping the `bin.first_unchecked` with a `std::panic::catch_unwind`. You will probably need to assert unwind safe because of the `Env`, so you'll probably need something like:

```suggestion
    #[should_panic]
    fn test_first_unchecked() {
        let env = Env::default();
        let mut bin = bytes![&env, [1, 2, 3, 4]];

        assert_eq!(bin.first_unchecked(), 1);
        bin.remove(0);
        assert_eq!(bin.first_unchecked(), 2);

        // first on empty bytes should panic
        let bin = bytes![&env];
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            bin.first_unchecked();
        }));
        assert_eq!(result, Err(...));
```

Reply:
Hi I am back, was traveling around so didn't have time to sit down. In terms of changes I changed the panic tests to catch the panic in a result variable (as you have shown here). Agree that it can be brittle as something else could trigger the panic. As for the test, I am unsure how I should test the type of panic (if necessary). I am already catching the panic and asserting that its indeed an error, as in this example:

```
fn test_first_unchecked() {
        let env = Env::default();
        let mut bin = bytes![&env, [1, 2, 3, 4]];
        assert_eq!(bin.first_unchecked(), 1);
        bin.remove(0);
        assert_eq!(bin.first_unchecked(), 2);

        // first on empty bytes should panic
        let bin = bytes![&env];
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            bin.first_unchecked();
        }));

        assert!(result.is_err())
    }
```

But can't find the correct way to test the type of panic. I was trying to follow how the Rust language does it ( https://github.com/rust-lang/rust/blob/a1bad57fa59a8069a6ebb05cd6a2ae73c88b2e98/src/test/run-pass/panic-uninitialized-zeroed.rs ), but this method is not working well at the point string equality is being tested.

I can commit what I have at the moment, if its evaluated as unnecessary to match the panic string message, I already changed all panic tests so as to follow the format I have posted above.

Reply:
I also deleted the redundant println messages found in the bytes file, as mentioned above.

Reply:
I agree, it's pretty gnarly. I managed to get it to work by it was too big:
```rust
assert_eq!(
    std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| bin.first_unchecked()))
        .map_err(|err| err
            .downcast_ref::<String>()
            .map(|s| { s.lines().into_iter().next().unwrap().to_owned() })),
    Err(Some("HostError: Error(Object, IndexBounds)".to_owned()))
);
```

In ef6c634348513525fc626d8cfbeec9c5ad31add9 I've tuned the should_panic use a bit so that the panic tests are smaller and focused on the one call.